### DISCUSSION TITLE
jlmkr on TrueNAS Dragonfish beta

### DISCUSSION THREAD

**Discussion by dheiss** [Jip-Hop/jailmaker#93] [General]
I just tested TrueNAS 24.04-BETA.1 where jlmkr throws an error when installing:
`Read-only file system: '/mnt/tank/jailmaker/jlmkr.py' -> '/usr/local/sbin/jlmkr'`
This was also the case when following the description in the TrueNAS manual for [sandboxes](https://www.truenas.com/docs/scale/scaletutorials/apps/sandboxes/). 

I was able to get around this by temporarily mounting /usr as rw:
`mount -o remount,rw /usr/`
`./jlmkr.py install`
`mount -o remount,ro /usr/`
Not perfect, but works for now.

**Comment by Jip-Hop** [comment]
Glad to hear you found a way around it but please have a look at this: 

https://github.com/Jip-Hop/jailmaker/issues/87#issuecomment-1943714580

**Comment by Jip-Hop** [comment]
The latest version of jailmaker should work out of the box on TrueNAS Dragonfish.

**Reply by janusn** [reply]
@Jip-Hop 
I have just upgraded my TrueNAS SCALE box to 24.04.0 and followed the installation instructions on the readme.md to install jlmkr again.

The command jlmkr.py install still throws an error of read only error as shown below:
```
# cd /mnt/tank/jailmaker/
# curl --location --remote-name https://raw.githubusercontent.com/Jip-Hop/jailmaker/main/jlmkr.py
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 70543  100 70543    0     0   326k      0 --:--:-- --:--:-- --:--:--  324k
# chmod +x jlmkr.py
# ./jlmkr.py --version
1.2.1
# ./jlmkr.py install
systemd-nspawn is already installed.
Cannot create symlink because /usr/local/sbin/jlmkr is on a readonly filesystem.
The bash alias jlmkr is already present.
The zsh alias jlmkr is already present.
Done installing jailmaker.
```
I have to remount /usr as read-write like @dheiss did in order to install jailmaker.

**Reply by Jip-Hop** [reply]
That is not an error. Just information. No need to remount rw at all. On 24.04 jlmkr works via a shell alias.

**Reply by janusn** [reply]
I was not so sure. It kept throwing errors like:
```
$ sudo systemctl status compose.service
Unit compose.service could not be found.
```

As least the jail compose was running fine after remounted as rw.

**Reply by Jip-Hop** [reply]
Refer to the main README. It explains about the alias and symlink on readonly filesystem.

**Reply by janusn** [reply]
@Jip-Hop 

I reinstalled TrueNAS SCALE 24.04.0 today and I install jailmaker 1.2.1 after configurations have been restored.

Without remounting /usr rw before installing jailmaker 1.2.1, `jlmkr` command only works with `sudo -s` not `sudo` as illustrated below.
```
$ sudo -s
# ./jlmkr.py install
systemd-nspawn is already installed.
Cannot create symlink because /usr/local/sbin/jlmkr is on a readonly filesystem.
The bash alias jlmkr is already present.
The zsh alias jlmkr is already present.
Done installing jailmaker.
# jlmkr list
NAME   RUNNING STARTUP GPU_INTEL GPU_NVIDIA OS     VERSION ADDRESSES
rocker False   True    False     False      debian 12      -    
# exit
exit
$ sudo jlmkr list
sudo: jlmkr: command not found
```

**Reply by Jip-Hop** [reply]
Ah interesting and thanks for sharing. I think the docs need clarification. I assume you first become root by running `sudo su`. All commands must be ran as root. This weekend is installed a fresh 24.04 VM with the root account disabled (the default option) to confirm this method still works. Let me know if you manage to get it working. A PR to update the README would be welcome too.

**Reply by janusn** [reply]
Thanks for your hard work @Jip-Hop 

I feel more confident running sudo than sudo -s as it is less error prone sometimes. 😝 But if it has to be the way forward, I am fine with it.

As for improving the docs, I would love to but it might not happen for awhile. As I am a newbie to github and I am busy tried to get the boot pool mirrored currently. That's why I reinstalled TrueNAS in the first place. 

And the reason making the mirror so difficult is, I have encountered weird issue that the second SSD can't be recognized. Then USB thumb drives are dead slow running TrueNAS SCALE. And those 512GB USB thumb drives I am trying are smaller than the native SSD installed. I have to buy a 1TB USB SSD in order to get it work. 😓
